src/main/java/org/folio/rest/impl/TenantPermsAPI.java

```diff
@@ -35,46 +44,395 @@
+          .onSuccess(v -> asyncResultHandler.handle(Future.succeededFuture(
+              PostTenantpermissionsResponse.respond201WithApplicationJson(entity))))
+          .onFailure(t -> {
+            logger.error(t.getMessage(), t);
+            asyncResultHandler.handle(Future.succeededFuture(
+                PostTenantpermissionsResponse.respond400WithTextPlain(t.getMessage())));
+          });
     } catch (Exception e) {
       logger.error(e.getMessage(), e);
       asyncResultHandler.handle(Future.succeededFuture(
           PostTenantpermissionsResponse.respond500WithTextPlain(e.getMessage())));
     }
   }
 
-  private Future<Void> savePermList(List<Perm> permList, Context vertxContext, String tenantId) {
+  /*
+   * Add permission.defined: "User" for any permission lacking the defined property. At this point
+   * all system-defined permissions should already have defined set.
+   */
+  private Future<Void> addMissingDefinedContext(Context vertxContext, String tenantId) {
+    Promise<Void> promise = Promise.promise();
+    Criteria crit = new Criteria().addField(DEFINED_FIELD).setOperation("IS NULL");
+
+    PostgresClient pgClient = PostgresClient.getInstance(vertxContext.owner(), tenantId);
+    pgClient.startTx(connection -> pgClient.get(TABLE_NAME_PERMS, Permission.class,
+        new Criterion(crit), true, false, getReply -> {
+          if (getReply.failed()) {
+            pgClient.rollbackTx(connection, done -> {
+              Throwable cause = getReply.cause();
+              logger.error(cause.getLocalizedMessage(), cause);
+              promise.fail(cause);
+            });
+            return;
+          }
+          List<Permission> userDefined = getReply.result().getResults();
+          userDefined.forEach(perm -> perm.setDefined(Defined.USER));
+          pgClient.upsertBatch(connection, TABLE_NAME_PERMS, userDefined, upsertReply -> {
+            if (upsertReply.failed()) {
+              pgClient.rollbackTx(connection, done -> {
+                Throwable cause = upsertReply.cause();
+                logger.error(cause.getLocalizedMessage(), cause);
+                promise.fail(cause);
+              });
+              return;
+            }
+            pgClient.endTx(connection, done -> {
+              logger.info(String.format("Added defined context to %s user-defined permissions",
+                  userDefined.size()));
+              promise.complete();
+            });
+          });
+        }));
+    return promise.future();
+  }
+
+  private Future<List<Permission>> getPermsForModule(ModuleId moduleId, List<OkapiPermission> perms,
+      Context vertxContext, String tenantId) {
+    return getPermsByModule(moduleId, vertxContext, tenantId)
+        .compose(existing -> {
+          if (existing.isEmpty() && perms != null) {
+            // this means either:
+            // A) the first time enabling this module, or
+            // B) the permissions exist but don't yet have the definedBy field
```

Comment:
The condition B) is the same as `the permissions exist but don't yet have the moduleName field`